tokio/src/time/driver/entry.rs

```diff
@@ -1,362 +1,616 @@
+        //
+        // This lock acquisition serves this purpose. All of the driver
+        // manipulations happen with the lock held, so we can just take the lock
+        // and be sure that this drop happens-after everything the driver did so
+        // far and happens-before everything the driver does in the future.
+        // While we have the lock held, we also go ahead and deregister the
+        // entry if necessary.
+        unsafe { self.driver.clear_entry(NonNull::from(self.inner())) };
+    }
 
-        self.waker.register_by_ref(cx.waker());
+    pub(crate) fn reset(mut self: Pin<&mut Self>, new_time: Instant) {
+        unsafe { self.as_mut().get_unchecked_mut() }.initial_deadline = None;
 
-        curr = self.state.load(SeqCst);
+        let tick = self.driver.time_source().deadline_to_tick(new_time);
 
-        if is_elapsed(curr) {
-            return Poll::Ready(if curr == ERROR {
-                Err(Error::from_u8(self.error.load(SeqCst)))
-            } else {
-                Ok(())
-            });
+        if let Ok(_) = self.inner().extend_expiration(tick) {
+            return;
         }
 
-        Poll::Pending
+        unsafe {
+            self.driver.reregister(tick, self.inner().into());
+        }
     }
 
-    /// Only called by `Registration`
-    pub(crate) fn reset(entry: &mut Arc<Entry>) {
-        let inner = match entry.upgrade_inner() {
-            Some(inner) => inner,
-            None => return,
-        };
-
-        let deadline = entry.time_ref().deadline;
-        let when = inner.normalize_deadline(deadline);
-        let elapsed = inner.elapsed();
-
-        let next = if when <= elapsed { ELAPSED } else { when };
+    pub(crate) fn poll_elapsed(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), super::Error>> {
+        if let Some(deadline) = self.initial_deadline {
+            self.as_mut().reset(deadline);
+        }
 
-        let mut curr = entry.state.load(SeqCst);
+        let this = unsafe { self.get_unchecked_mut() };
 
-        loop {
-            // In these two cases, there is no work to do when resetting the
-            // timer. If the `Entry` is in an error state, then it cannot be
-            // used anymore. If resetting the entry to the current value, then
-            // the reset is a noop.
-            if curr == ERROR || curr == when {
-                return;
-            }
+        this.inner().state.poll(cx.waker())
+    }
+}
 
-            let actual = entry.state.compare_and_swap(curr, next, SeqCst);
+impl TimerHandle {
+    pub(super) unsafe fn cached_when(&self) -> u64 {
+        unsafe { self.inner.as_ref().cached_when() }
+    }
 
-            if curr == actual {
-                break;
-            }
+    pub(super) unsafe fn sync_when(&self) -> u64 {
+        unsafe { self.inner.as_ref().sync_when() }
+    }
 
-            curr = actual;
-        }
+    // used from a debug_assert! only
+    #[allow(dead_code)]
```

Comment:
TIOLI: could just be
```suggestion
    // used from a debug_assert! only
    #[cfg(debug_assertions)]
```

Reply:
Actually looks like this is really unused now, so I'll just delete it.